components/brave_wallet/browser/solana_compiled_instruction.cc

```diff
@@ -0,0 +1,153 @@
+#include "brave/components/brave_wallet/browser/solana_message_address_table_lookup.h"
+#include "brave/components/brave_wallet/common/solana_address.h"
+#include "brave/components/brave_wallet/common/solana_utils.h"
+
+namespace brave_wallet {
+
+namespace {
+
+absl::optional<uint8_t> FindIndexInStaticAccounts(
+    const std::vector<SolanaAddress>& keys,
+    const std::string& target_key) {
+  if (keys.size() > UINT8_MAX) {
+    return absl::nullopt;
+  }
+  auto it = base::ranges::find_if(keys, [&](const SolanaAddress& key) {
+    return key.ToBase58() == target_key;
+  });
+  if (it == keys.end()) {
+    return absl::nullopt;
+  }
+  return it - keys.begin();
+}
+
+absl::optional<uint8_t> FindIndexInAddressTableLookups(
+    const std::vector<SolanaMessageAddressTableLookup>& addr_table_lookups,
+    const SolanaAccountMeta& account,
+    uint8_t num_of_static_accounts,
+    uint8_t num_of_write_indexes,
+    uint8_t num_of_read_indexes) {
+  uint8_t index_of_combined_array =
+      account.is_writable ? num_of_static_accounts
+                          : num_of_static_accounts + num_of_write_indexes;
+
+  for (const auto& addr_table_lookup : addr_table_lookups) {
+    const auto& indexes = account.is_writable
+                              ? addr_table_lookup.write_indexes()
+                              : addr_table_lookup.read_indexes();
+    if (account.pubkey != addr_table_lookup.account_key().ToBase58()) {
+      index_of_combined_array += indexes.size();
+      continue;
+    }
+
+    for (size_t i = 0; i < indexes.size(); ++i) {
+      if (*account.address_table_lookup_index == indexes[i]) {
+        return index_of_combined_array + i;
+      }
+    }
+  }
+
+  return absl::nullopt;
+}
+
+}  // namespace
+
+SolanaCompiledInstruction::SolanaCompiledInstruction(
+    uint8_t program_id_index,
+    const std::vector<uint8_t>& account_indexes,
+    const std::vector<uint8_t>& data)
+    : program_id_index_(program_id_index),
+      account_indexes_(account_indexes),
+      data_(data) {}
+
+SolanaCompiledInstruction::SolanaCompiledInstruction(
+    SolanaCompiledInstruction&&) = default;
+SolanaCompiledInstruction& SolanaCompiledInstruction::operator=(
+    SolanaCompiledInstruction&&) = default;
+SolanaCompiledInstruction::~SolanaCompiledInstruction() = default;
+
+bool SolanaCompiledInstruction::operator==(
+    const SolanaCompiledInstruction& ins) const {
+  return program_id_index_ == ins.program_id_index_ &&
+         account_indexes_ == ins.account_indexes_ && data_ == ins.data_;
+}
+
+// static
+absl::optional<SolanaCompiledInstruction>
+SolanaCompiledInstruction::FromInstruction(
+    const SolanaInstruction& instruction,
+    const std::vector<SolanaAddress>& static_accounts,
+    const std::vector<SolanaMessageAddressTableLookup>& addr_table_lookups,
+    uint8_t num_of_write_indexes,
+    uint8_t num_of_read_indexes) {
+  auto program_id_index =
+      FindIndexInStaticAccounts(static_accounts, instruction.GetProgramId());
```

Comment:
Can we add a comment saying why program id must came from static accounts?

Reply:
46f9f027bd45d2f03324c63a9724ddbbe1812de0